### ISSUE TITLE
[HOLD for payment 2023-10-16] [$500] Assign Task - Assignee field active in offline mode while assigning task, should be greyed out

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#25956]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat
2. Navigate to the "Assign task" section
3. Create a task by entering a title
4. Enable offline mode from Preferences
6. Select a user from the assignee field
7. Notice assignee field is not greyed out

## Expected Result:
During offline mode, the assignee field should appear greyed out to indicate its inactivity while assigning a task

## Actual Result:
The assignee field remains active and doesn't turn grey while assigning a task in offline mode

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] Windows / Chrome 
- [ ] MacOS / Desktop

**Version Number:** 1.3.57-4

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/6116d6f9-8352-4a2d-abf6-6c3a97791687

https://github.com/Expensify/App/assets/78819774/dd482945-027e-4ed7-9c97-a584289095de

**Expensify/Expensify Issue URL:**

**Issue reported by:** @ayazhussain79

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691792826667939

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01cb180f9f96ed4319</li>
        <li>Upwork Job ID: 1697586364732370944</li>
        <li>Last Price Increase: 2023-09-15</li>
    </ul>
</details>

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Assignee field active in offline mode while assigning task, should be greyed out

### What is the root cause of that problem?
We don't set pendingField when editing assignee 
https://github.com/Expensify/App/blob/c435734d8cf360e964e20eea1fb54c541f6fa7ea/src/libs/actions/Task.js#L511
### What changes do you think we should make in order to solve the problem?
In optimistic report data
https://github.com/Expensify/App/blob/c435734d8cf360e964e20eea1fb54c541f6fa7ea/src/libs/actions/Task.js#L526-L531
we should add pending field managerID: update like this

```
pendingFields: {
        ...(assigneeAccountID && {managerID: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE}),
},
```
We also need to remove pendingFields when updating successful
### What alternative solutions did you explore? (Optional)

**Comment by esh-g** [comment]
# Proposal

## Please re-state the problem we are trying to solve
The task assignee field is doesn't get greyed out when changed offline.

## What is the root cause of this problem?
This was introduced in this PR: https://github.com/Expensify/App/pull/22778
We are not setting the `pendingField` for `managerID` while editing task assignee in optimistic data
https://github.com/Expensify/App/blob/a03a21d25832760cf9db4235b0fbc88cc08bd490/src/libs/actions/Task.js#L525-L534

Therefore, the pending field for `managerID` is always `null` and it never gets greyed out.

## What changes do you think we should make in order to solve the problem?
1. We should add the `pendingField` to the optimistic data while editing task assignee like this:
```js
{
    onyxMethod: Onyx.METHOD.MERGE,
    key: `${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`,
    value: {
        reportName,
        managerID: assigneeAccountID || report.managerID,
        managerEmail: assigneeEmail || report.managerEmail,
        pendingFields: {managerID: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE}, // add this line
    },
},
```
2. In success data, we should add the following to clear out the pending fields
```js
{
    onyxMethod: Onyx.METHOD.MERGE,
    key: `${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`,
    value: {pendingFields: {managerID: null}}
}
```

This is the same approach we follow `editTaskAndNavigate` function.

**Comment by michaelhaxhiu** [comment]
@allroundexperts - do you think this bug report & problem is a dupe of https://github.com/Expensify/App/issues/23200? Would *love* your input!

**Comment by ayazhussain79** [comment]
@michaelhaxhiu This bug https://github.com/Expensify/App/issues/23200 [PR](https://github.com/Expensify/App/issues/23200) merged but https://github.com/Expensify/App/issues/25956 is still reproducible @allroundexperts can you please also check this one

**Comment by michaelhaxhiu** [comment]
Hmm let's see what's suggested on https://github.com/Expensify/App/issues/23200 from those involved. You may be due a bug reporting bounty if this is indeed "separate" and thus unique.

**Comment by michaelhaxhiu** [comment]
But firstly I want to ensure this is fixed. If it's not fixed, there is no bug bounty to pay. That's the pre-requisite.

**Comment by michaelhaxhiu** [comment]
https://github.com/Expensify/App/pull/22778#issuecomment-1701261268

Commented here for team input on how to best proceed. This is a valid bug and still occurring based on latest testing, so we *do need to fix it*. But I want to check if we should draft a new PR or amend/update the PR that caused it.

**Comment by michaelhaxhiu** [comment]
cc'ing 
@thienlnam
@Santhosh-Sellavel

**Comment by michaelhaxhiu** [comment]
this is a regression from https://github.com/Expensify/App/pull/22778

**Comment by michaelhaxhiu** [comment]
might make sense for @abdulrahuman5196 to trade this GH to @Santhosh-Sellavel since he was involved in the parent regression, though it's not necessary. I'll let yall decide.

**Comment by Rachit-pul** [comment]
Hello Team, here from the upwork job link.

After having a cursory look and watching the video, this is the proposal I could come up with. If you think it is worth exploring please contact me and I'll happily set up a meeting to discuss further and clarify some doubts

Proposal


**Please re-state the problem we are trying to solve**

The task assignee field is doesn't get greyed out when changed offline.

**What is the root cause of this problem?**

This seems like a simple rendering issue where in the field does not have the logic to be greyed out. I am assuming in the previous fixed the forced online variable is accessible here.

Two scenarios I could come up with,
1. If force_offline is already true, before the user reaches here. The field is greyed out. In this case the new change from false -> true is not synced with the switch
2. It is never greyed out regardless of the switch

**What changes do you think we should make in order to solve the problem?**

In both cases, the solution would simply be in 2 parts, 

1. Adding the logic to check force_offline before rendering.
2. On every switch between true/false we update the following page. This would require a little deeper understanding of the codebase to see how these chat instances are stored and if the GUI objects are available post rendering. 

Also, in case the force_offline is a debugging/testing tool. I can make it so that it checks actually being connected to the internet and then switch between greyed out or not.( Should work for most modern browsers)


This was just a cursory look, in case I am off by a lot please guide me a little as to how the implementation works and I can come up with a detailed report on the fix, But from the problem statement and video this is what I could predict.

Thanks!

Rachit Pulhani
https://www.upwork.com/freelancers/~01167d06dbad3ad324
rachit2510pulhani@gmail.com

**Comment by akdevndesign** [comment]
Proposal

**Please re-state the problem we are trying to solve**
The task assignee field doesn't get greyed out (or block user from accessing assignee menu) while offline.

**What is the root cause of this problem?**
There is  no existing logic to update state of assignee field when online status is set to offline. 

**What changes do you think we should make in order to solve the problem?**
 To begin, I would modify NewTaskPage component. I would create conditional styles to apply to the "task assignee". Looks like the is a enabledWhenOffline prop already used in the same component - so I would check if it is compatible with the "task asignee" field.  Depending on the functionality of the enabledWhenOffline prop, I might have to create a custom style to handle the opacity changes while offline. 
 
 Thanks! 
 
 Aaron K.

**Comment by Rachit-pul** [comment]
Contributor details
Your Expensify account email: rachit2510pulhani@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01167d06dbad3ad324

**Comment by akdevndesign** [comment]
Contributor details
Your Expensify account email: [ajkelsey53@gmail.com](mailto:ajkelsey53@gmail.com)
Upwork Profile Link: (https://www.upwork.com/freelancers/aaronk10)

**Comment by akdevndesign** [comment]
Contributor details
Your Expensify account email: ajkelsey53@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/aaronk10

**Comment by abdulrahuman5196** [comment]
I can review this issue. The linked regression bug is like a month old so this should be fine.

**Comment by michaelhaxhiu** [comment]
Waiting for review by @abdulrahuman5196. Let us know when you anticipate having a chance to prioritize!

**Comment by abdulrahuman5196** [comment]
Sorry for the delay. Will cover over the weekend

**Comment by michaelhaxhiu** [comment]
Bump, any update here?

**Comment by akdevndesign** [comment]
I know this isn't the process but if you still need someone to squash this - still happy to assist!  I can knock out ASAP.

**Comment by abdulrahuman5196** [comment]
TAL now

**Comment by abdulrahuman5196** [comment]
Proposal from @DylanDylann here https://github.com/Expensify/App/issues/25956#issuecomment-1693608602 looks good and works well.

🎀 👀 🎀
C+ Reviewed

**Comment by michaelhaxhiu** [comment]
I'm re-assigning this to another BZ as part of my preparation for Sabbatical (starting Friday).

**Next steps:**

- Sherpa this forward with @abdulrahuman5196
- After the PR is finalized, pay the bug reporter / contributor / C+ (based on timestamp, the payouts will honor our old price points btw [context](https://expensify.slack.com/archives/C01SKUP7QR0/p1695893416991979))

**Comment by abdulrahuman5196** [comment]
Overdue is Pending on @tylerkaraszewski review.

**Comment by DylanDylann** [comment]
@abdulrahuman5196 The PR is ready for review: https://github.com/Expensify/App/pull/28378

**Comment by abekkala** [comment]
### PAYMENTS FOR OCT 16

- @ayazhussain79 (Reporter) [$250] (reported prior to price change) - Upwork offer sent
- @DylanDylann (Contributor) [$500 - no bonus] - Upwork offer sent
- @abdulrahuman5196 (Reviewer) [$500 - no bonus] - Upwork offer sent

**Comment by DylanDylann** [comment]
@abekkala Could you help to check timeline bonus again? The PR is merged in 3 days

**Comment by abekkala** [comment]
@DylanDylann and @abdulrahuman5196 ah, yes - I do see that Melvin missed this one. 

@abdulrahuman5196 please complete checklist prior to payment

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/22778/files

> The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

https://github.com/Expensify/App/pull/22778/files#r1361115083

>Determine if we should create a regression test for this bug.

Yes.

>If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.


1) Create a task
2) Enable offline mode from Preferences
3) Update the task assignee
4) Verify assignee field is greyed out
5) Disable offline mode from Preferences
6) Verify all field is updated


@abekkala

**Comment by abekkala** [comment]
@ayazhussain79 paid and contract ended - thank you! 🎉 

@DylanDylann paid and contract ended - thank you!  🎉 

@abdulrahuman5196 paid and contract ended - thank you! 🎉
